### DISCUSSION TITLE
How do I   understand points in raster cell center?

### DISCUSSION THREAD

**Discussion by zhengliuX** [imr-framework/pypulseq#160] [Q&A] [answered]
Hi, when reading specification section 2.6, it says 
```
In contrast to that, any default sampling points (e.g. sampling points of arbitrary-shape gradients) are aligned to the centers of the corresponding raster steps.
```
Which means for **RF**, **Arbitrary Gradient** and **ADC**, when you are designing the these events, I should know that the points I designed will be in the center of a raster time cell.  

I've read https://github.com/pulseq/pulseq/issues/32#issuecomment-1707839354, if I understand that discussion right, when designing an ADC event, the real points in the center of raster time cell are the accurate I want the points to be, no matter how the real hardware designs, as for hardware may need some warm-up time, causing a little delay. Is this the same way to understand my RF and arbitrary gradients points? The accurate time I want the points generated by hardware. 

So if I set RF delay as rf_delay, **the first point** should be 
```
rf_delay + 0.5 * rf_dwelltime
```
same for arbitrary gradients 
```
grad_delay + 0.5 * grad_dwelltime
```
and that is the accurate time the point will be. To do a tricky thing, if I want events to start at rf_delay(for RF), in my code, I should set delay as 
```
rf_delay - 0.5 * rf_dwelltime
```
Then Pulseq will set it to 
```
rf_delay - 0.5 * rf_dwelltime + 0.5 * rf_dwelltime = rf_delay
```

To a Pulseq interpreter, when parsing a .seq file, it gets the accurate time of the points. Let's say it gets the first point start time is `rf_delay + 0.5 * rf_dwelltime`, which is calculated using `rf.delay + rf.t[0]`, then the interpreter just makes sure the first point occurring at `rf_delay + 0.5 * rf_dwelltime` second relative to the start time of the block.

--- 

Also, for hard RF pulseq(or block pulseq based on https://github.com/pulseq/pulseq/blob/master/matlab/%2Bmr/makeBlockPulse.m#L57 and https://github.com/imr-framework/pypulseq/blob/dev/pypulseq/make_block_pulse.py#L105), 
it's defined on the edge, different from other RF pulses, is this excepted?

Thank a lot!
Zheng

**Comment by btasdelen** [comment]
Hi @zhengliuer, I am not the best person to answer this, but no one chimed in, so take this with a grain of salt.

To my understanding, that half dwell time shift is done by the ~~Scanner, (Py)Pulseq does the shift for you to know what the actual waveform is (be it gradient, RF or ADC), not that it saves shifted waveforms to .seq file~~ (Apparently, it is stored shifted in .seq file). We still define all the pulses, including block pulses, by their edges, not mid-points.

However, when we are designing the waveforms, especially the gradients, we have to know that they will be played by their mid-raster points, because that is what will determine the trajectory (or excitation k-space for RF).

I hope someone will correct me if this is inaccurate.

**Reply by zhengliuX** [reply]
So there are two parts:
1. sequence designer
For the designers, people should know that although I designed an arb grad starting at 0, in a real scanner, it starts at 0.5. 
3. interpreter
When an interpreter reads a .seq file, it gets the delay and dwell time, and does the time shifting based on the type by itself like
```
start_time = delay + 0.5*dwell
```
---
However, what confuses me is this:
In this epi demo, https://github.com/imr-framework/pypulseq/blob/dev/pypulseq/seq_examples/scripts/write_epi.py#L58-L62
to align with the gradient, it sets adc delay as gx.rise_time - 0.5*dwell, for gx is a trapezoid gradient which starts at the edge, or simply, 0. But adc is a type of "waveform", to the interpreter will shift its time 0.5*dwell, by minus 0.5*dwell_time, it will align with the gx's flattop.

But in this gre demo, https://github.com/imr-framework/pypulseq/blob/dev/pypulseq/seq_examples/scripts/write_gre.py#L51-L53
it just sets adc delay as gx.rise_time, so in a real scanner, it will start at 0.5*dwell, which causes unalignment.

Did I understand these wrong or I miss something?

**Reply by btasdelen** [reply]
See the answer from @schuenke, apparently the shift is done by Pulseq, not the scanner.

I think the reason why `write_epi` shifts by half dwell time is to skip the center of the k-space and make the k-space symmetrical. So normally we would acquire `k_x` from -16 to 15, but EPI has bipolar readout gradients. To avoid acquiring 16 to -15 during the negative gradient, they acquire -15.5 to 15.5.

**Reply by zhengliuX** [reply]
Oh yes, skipping the center and making it symmetrical makes more sense. Thanks a lot!

**Comment by schuenke** [comment]
Regarding the waveform shift: the shift IS stored in the seq-file. If you define an RF pulse with duration 1 ms and 100 samples, the first value stored in the seq-file is the value at 5 µs, the second at 15 µs and so on...

Regarding the ADC: a single adc point is not an instantaneous acquisition. You acquire data over a timeframe with duration dwell. So IMO an ADC, where the first sample point is given at dwell/2, will still start at 0. The first "point" corresponds to the time frame from 0 to dwell.

**Reply by zhengliuX** [reply]
Thank you so much for the reply!
1. __time shift__:
From an interpreter's perspective, according to the .h file, it seems that no api for get parameters about shifting? From what I can see now is that I can get dwelltime, delay and signal magnitude vector. How is that stored in a .seq file? The interpreter will directly read that?

3. __ADC:__
So if I set an ADC point as start_time=0.5*dwell, its real time range is
```
[start_time-0.5*dwell, start_time+0.5*dwell]
```
(This I understand for it's analog). As for why is 0.5, not other numbers, that's from Siemens's specification?


Also, when reading the spiral demo, [this](https://github.com/pulseq/tutorials/blob/main/12_Radial_and_nonCartesian/seq/s06_Spiral.m#L77-L94) says` Siemens will try to split the trajectory into segments <=1000 samples`, so does this mean I can design a very large number of adc points?

**Reply by schuenke** [reply]
you can design ADCs with more than 1000 points, but, like you cited, the Siemens interpreter will try to split a single huge ADC in smaller ADS with <= 1000 samples. As far as I remember, all samples need to have the same size. Maybe check the documentation again if you want to use such long ADCs with arbitrary number of samples. Or simply use multiple of 100 or 1000 as ADC points.

**Comment by zhengliuX** [comment]
Hi, an additional question. Like ADC, when I set ADC points in the raster cell center, the time frame is from t-dwell/2 to t+dwell/2, is this the same as arbitrary gradient and RF?
If I set an RF point at 0, it is expected as in the center of raster cell 0.5 according to the specification, but the time frame of this point is still 0 ~ 1? Like Its expected to hold the value in t_rf-dwell/2 to t_rf+dwell/2? Thanks!
@btasdelen @schuenke

**Reply by schuenke** [reply]
Yes, correct @zhengliuer. Imagine you would not hold the value of and arb. gradient. This would mean you had to switch it on/off and ramp to the desired amp within a dwell time. Due to limited slew rates etc this is technically not even possible.

**Reply by zhengliuX** [reply]
Thanks a lot @schuenke ! But just a little additional info for my confusion. As I know, some equipments will hold the value of the point after one dwell time of the start time of it. But in Pulseq, it seems I should understand that the hardware will hold the value dwell/2 before the point(which is set as the center).
![12141e03-c0c9-4082-894c-5744f67b3e2e](https://github.com/imr-framework/pypulseq/assets/65115609/16aec5d0-f97b-46d9-a66b-48d2cc6b2191)
Like the figure shows above, the left means the actual time frame that hardware holding the value is dwell/2 before and after the point, the the right means the hardware holding the value one dwell after the point.

**Reply by schuenke** [reply, marked as answer]
The left version should be correct. 

If you rephrase it to: the value that is hold during the first dwell intervall is defined by the value at the center of the intervall [0 dwell], it sounds much more intuitive than saying "pulseq will hold the value dwell/2 before the point" IMO

**Reply by zhengliuX** [reply]
Thanks a lot! And thank you for the rephrasing, it is really clearer!
